### PR TITLE
Ensure CSV exports write the header row even if there are no entries. Clean up internal implementation.

### PR DESCRIPTION
**Pull request by clouser2**
### Description
Noticed as part of #1743. The interface for exporting CSVs didn't have clear separation of concerns or object lifecycle. We had logic to write the header row when writing the first record. While here, I took the opportunity to clean up the interface a bit:
  * Removed ExporterFactory, since it only has a single callsite
  * A `CSVExporter` wraps a `CSVPrinter` and implements `AutoClosable`. This means that the header row is written as part of the constructor (similar to the interface of `CSVPrinter`). This also means that we no longer have to pass a `Writer` to every invocation of a record, hopefully making the lifecycle clearer. 

### Checklist
- [X] Created tests which fail without the change (if possible)

### Issue(s)
 #1743

### CODE REVIEW

**Comment by clouser2** [comment]
Note: I expect the formatter to fail since it presently has a bug where it runs on deleted files. Sending a PR out for that shortly.

File: server/app/services/export/CsvExporter.java [hunk 1]
```diff
@@ -22,59 +21,39 @@
  * {@link Path} indexing into an applicant's data, and CsvExporter takes the path and reads the
  * answer from {@link ReadOnlyApplicantProgramService} if present.
  */
-public class CsvExporter {
-  public static final CSVFormat DEFAULT_CSV_FORMAT =
```
Comment (clouser2) [inline server/app/services/export/CsvExporter.java:26]: Note to the reviewer: This was unnecessarily shared between the implementation and the test.

File: server/app/services/export/CsvExporter.java [hunk 2]
```diff
@@ -22,59 +21,39 @@
...
-      CSVFormat.DEFAULT.builder().setHeader().setSkipHeaderRecord(true).build();
-
+public final class CsvExporter implements AutoCloseable {
```
Comment (bion) [inline server/app/services/export/CsvExporter.java:24]: suggestion (blocking): update javadoc to mention the need to close

File: server/app/services/export/CsvExporter.java [hunk 3]
```diff
@@ -22,59 +21,39 @@
...
+    this.printer = new CSVPrinter(writer, format);
   }
 
   /**
-   * The CSV exporter will write the headers on first call to services.export(). It does not store
-   * the writer between calls. Since it is intended for many applications, this function is intended
-   * to be called several times.
+   * The CSV exporter does not store the writer between calls. Since it is intended for many
+   * applications, this function is intended to be called several times.
    */
```
Comment (bion) [inline server/app/services/export/CsvExporter.java:48]: suggestion (blocking): the comment for this method beginning with "The CSV exporter does not store the writer between calls." is a little strange now that the method doesn't take a writer. Please reword or remove

File: server/app/services/export/ExporterService.java [hunk 4]
```diff
@@ -127,49 +127,49 @@ private CsvExportConfig generateDefaultCsvExportConfig(
...
-        .anyMatch(exportDefinition -> exportDefinition.csvConfig().isPresent())) {
-      csvExporter = exporterFactory.csvExporter(program.toProgram());
-    } else {
-      csvExporter = exporterFactory.csvExporter(generateDefaultCsvConfig(programId));
-    }
+    Optional<CsvExportConfig> maybeExportConfig =
+        program.exportDefinitions().stream()
+            .filter(exportDefinition -> exportDefinition.csvConfig().isPresent())
+            .map(exportDefinition -> exportDefinition.csvConfig().get())
+            .findFirst();
```
Comment (bion) [inline server/app/services/export/ExporterService.java:134]: suggestion (non-blocking): i'm pretty sure that admin-specified, per-program export configuration was a planned feature that never got implemented. could remove this logic altogether if we wanted i think, though that may be refactor outside the scope of this PR

**Review by bion — CHANGES_REQUESTED** [review]

**Review by bion — APPROVED** [review]

File: server/app/services/export/CsvExporter.java [hunk 5]
```diff
@@ -21,60 +21,40 @@
  * CsvExporter takes a list of {@link Column}s and exports the data specified. A column contains a
  * {@link Path} indexing into an applicant's data, and CsvExporter takes the path and reads the
  * answer from {@link ReadOnlyApplicantProgramService} if present.
+ *
+ * <p>The close method should be called when complete in order for the underlying {@link CsvPrinter}
```
Comment (kmcnellis) [inline server/app/services/export/CsvExporter.java:25]: According to the java docs, close is called automatically if you have a try-with-resources.  This comment seems a bit incorrect.  Could you modify it to say you need to either close() or  try-with-resources?

**Review by kmcnellis — APPROVED** [review]

**Pull request merged**